design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+```
+
+Given data in the pageviews topic:
+```sql
+3,,
+3,user,page
+2,,page1
+null,user,page
+```
+
+A `SELECT` query from the table would return only the two records that didn't violate the `NOT NULL` constraint:
+```sql
+ksql> select * from pageviews emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+|null           |user                                                            |page1                                                           |
+```
+
+Declaring a column as `NOT NULL` in a CSAS/CTAS would look like:
+```sql
+CREATE stream pageviews_no_null as select CAST(viewtime AS BIGINT NOT NULL), userid, pageid from pageviews;
+```
+The resulting stream's columns would all be `NOT NULL`
+
+```sql
+ksql> select * from pageviews_no_null emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+```
+
+Attempting to insert invalid records would error out:
+```sql
+ksql> INSERT INTO pageviews VALUES (5, NULL, 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "USERID"
+
+ksql> INSERT INTO pageviews VALUES (NULL, 'user3', 'page23'); # this insert succeeds
+
+ksql> INSERT INTO pageviews_no_null VALUES (NULL, 'user3', 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "VIEWTIME"
+```
+
+`NOT NULL` for arrays, maps, struct
+```sql
+CREATE STREAM valid_data (arr ARRAY<BIGINT NOT NULL>, mapping MAP<VARCHAR NOT NULL, BIGINT NOT NULL>, valid_struct STRUCT<c DOUBLE NOT NULL, d BOOLEAN>) ...
+```
+Records that have column values such as
+* `{arr: [3,4,null] ... }`
+* `{mapping: {"key": null} ...}`
+* `{mapping: {null: 5} ...}`
+* `{valid_struct: {c: null, d: true} ...}`
+
+Would be dropped when reading from the `valid_data` stream and trying to insert record with these column values would error out.
+
+## NOT NULL DEFAULT
```

Comment:
Would the DEFAULT work only along NOT NULL? What about when inserting data and not specifying a column that has a default value, would that use the default value?

Reply:
Example:
```
ksql> create stream s1 (id int, name string default 'unamed') with ...
ksql> insert into s1(id) values (1);
ksql> select * from s1 emit changes:
id |  name
1  |  'unamed'
```

I know DEFAULT is not part of the scope, but wouldn't be easier to handle NOT NULL and DEFAULT separate than forcing to use them together?

Reply:
+1. I also feel it's better to decouple the NOT NULL from DEFAULT syntax. For this specific use case of `ALTER`, I'm a bit leaning towards not using ALTER to modify the semantics constraint at the moment; and if we do want to do this now, maybe we should use the syntax like `ADD CONSTRAINT`.